database/rrdcontext.c

```diff
@@ -0,0 +1,2543 @@
+                   ",\n\t\t\t\"first_time_t\":%ld"
+                   ",\n\t\t\t\"last_time_t\":%ld"
+                   ",\n\t\t\t\"collected\":%s"
+                   ",\n\t\t\t\"deleted\":%s"
+                   , rc->version
+                   , rc->hub.version
+                   , string2str(rc->title)
+                   , string2str(rc->units)
+                   , string2str(rc->family)
+                   , rrdset_type_name(rc->chart_type)
+                   , rc->priority
+                   , rc->first_time_t
+                   , rrd_flag_is_collected(rc) ? t->now : rc->last_time_t
+                   , rc->flags & RRD_FLAG_COLLECTED ? "true" : "false"
+                   , rc->flags & RRD_FLAG_DELETED ? "true" : "false"
+                   );
+
+    if(options & RRDCONTEXT_OPTION_SHOW_FLAGS) {
+        buffer_strcat(wb, ",\n\t\t\t\"flags\":\"");
+        rrd_flags_to_buffer(rc->flags, wb);
+        buffer_strcat(wb, "\"");
+    }
+
+    if(options & RRDCONTEXT_OPTION_SHOW_QUEUED) {
+        buffer_strcat(wb, ",\n\t\t\t\"queued_reasons\":\"");
+        rrd_reasons_to_buffer(rc->queue.queued_flags, wb);
+        buffer_strcat(wb, "\"");
+
+        buffer_sprintf(wb,
+                       ",\n\t\t\t\"last_queued\":%llu"
+                       ",\n\t\t\t\"scheduled_dispatch\":%llu"
+                       ",\n\t\t\t\"last_dequeued\":%llu"
+                       , rc->queue.queued_ut / USEC_PER_SEC
+                       , rc->queue.scheduled_dispatch_ut / USEC_PER_SEC
+                       , rc->queue.dequeued_ut / USEC_PER_SEC
+                       );
+    }
+
+    if(options & (RRDCONTEXT_OPTION_SHOW_INSTANCES|RRDCONTEXT_OPTION_SHOW_METRICS)) {
+        buffer_sprintf(wb, ",\n\t\t\t\"charts\": {");
+        struct rrdcontext_to_json tt = {
+            .wb = wb,
+            .options = options,
+            .written = 0,
+            .now = t->now,
+        };
+        dictionary_sorted_walkthrough_read(rc->rrdinstances, rrdinstance_to_json_callback, &tt);
+        buffer_strcat(wb, "\n\t\t\t}");
+    }
+
+    buffer_strcat(wb, "\n\t\t}");
+    t->written++;
+    return 1;
+}
+
+int rrdcontext_to_json(RRDHOST *host, BUFFER *wb, RRDCONTEXT_TO_JSON_OPTIONS options, const char *context) {
+    RRDCONTEXT_ACQUIRED *rca = (RRDCONTEXT_ACQUIRED *)dictionary_get_and_acquire_item((DICTIONARY *)host->rrdctx, context);
+    if(!rca) return HTTP_RESP_NOT_FOUND;
+
+    RRDCONTEXT *rc = rrdcontext_acquired_value(rca);
+
+    if(options & RRDCONTEXT_OPTION_DEEPSCAN)
+        rrdcontext_recalculate_context_retention(rc, RRD_FLAG_NONE, -1);
+
+    struct rrdcontext_to_json t = {
+        .wb = wb,
+        .options = options|RRDCONTEXT_OPTION_SKIP_ID,
+        .written = 0,
+        .now = now_realtime_sec(),
+    };
+    rrdcontext_to_json_callback(context, rc, &t);
+
+    rrdcontext_release(rca);
+    return HTTP_RESP_OK;
+}
+
+int rrdcontexts_to_json(RRDHOST *host, BUFFER *wb, RRDCONTEXT_TO_JSON_OPTIONS options) {
```

Comment:
I am of opinion all our JSON handling should use JSONC instead of handcrafted strings